add Cyclone V SoC FPGA support

mevalver opened the pull request:
soc: arm: privilege: add Cyclone V SoC FPGA support

Add support for the Intel Cyclone V SoC FPGA (arm Cortex-A9).

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

mevalver committed:
soc: arm: Add Cyclone V SoC FPGA Support

soc: arm: privilege: add Cyclone V SoC FPGA suppport
Add support for the Intel Cyclone V SoC FPGA (arm Cortex-A9).

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

mevalver committed:
boards: arm: add basic Cyclone V SoC FPGA board DK definition

Add a basic board definition for the Cyclone V SoC FPGA.

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

mevalver committed:
scripts: west_commands: runners: add runner for Cyclone V SoC FPGA DK

Add a runner to "flash" and "debug" Cyclone V SoC FPGA Development Kit
the runner is based on OpenOCD and GDB

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

mevalver committed:
samples: drivers: adding sample to use Cyclone V SoC FPGA DK LCD display

LCD display is connected to the I2C bus SoC bus in the development kit,
this sample guides the user on how to use the LCD display with I2C commands

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

mevalver committed:
drivers: serial: modify ns16550 to use extended FIFO

Cyclone V SoC FPGA supports 128Byte FIFO for UART communication,
this modification adds a feature to use 128byte FIFO serial UART

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

carlescufi commented on dts/arm/intel_socfpga_std/socfpga.dtsi:
```diff
@@ -0,0 +1,300 @@
+ * Copyright (C) 2022, Intel Corporation
+ * Description:
+ * Section of the DTS file containing definitions for the HPS
+ * based on Linux, heavily modified for Zephyr
```
Just checking, if this is based on Linux, did you relicense it from GPLv2 to Apache v2?

mevalver replied:
We created the DTS based on our Golden System Reference Design DTS... The file is similar, as the addresses are the same, but of course, we built this one for Zephyr.

carlescufi commented on soc/arm/intel_socfpga_std/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+# Description:
+# Adding support for Intel Cyclone V SoC
+
+add_subdirectory(${SOC_SERIES})
```
Please add newlines at the end of all files. All editors have an option to do that.

carlescufi commented:
@mevalver thanks for the PR! Please do not close it and open a new one if you need to make changes. Everything in the PR can be edited/modified without the need for creating a new one.

nashif commented:
> Signed-off-by: mevalver <esteban.valverde.vega@intel.com>

@mevalver this should be your full name, please see https://docs.zephyrproject.org/latest/contribute/guidelines.html#signed-off-by


git config --global user.name "David Developer"
git config --global user.email "david.developer@company.com"

nashif commented on boards/arm/cyclonev_socdk/cyclonev_socdk_defconfig:
```diff
@@ -0,0 +1,63 @@
+
+# I2C:
+CONFIG_I2C=y
+CONFIG_I2C_DW=y
```
this can be added in the Kconfig.defconfig file

if I2C
config I2C_DW
	default y
endif /* I2C */


and for all others below...

mevalver replied:
This is solved

nashif commented on drivers/serial/uart_ns16550.c:
```diff
@@ -89,12 +89,15 @@ BUILD_ASSERT(IS_ENABLED(CONFIG_PCIE), "NS16550(s) in DT need CONFIG_PCIE");
 #define FCR_RCVRCLR 0x02 /* clear RCVR FIFO */
 #define FCR_XMITCLR 0x04 /* clear XMIT FIFO */
+#define FCR_TET_MASK 0x30 /* Empty Threshold level mask */
+#define FCR_RT_MASK 0xC0 /* Rx Trigger level mask */
```
where are you using those?

carlocaione commented on drivers/i2c/i2c_dw.c:
```diff
@@ -42,7 +42,12 @@ LOG_MODULE_REGISTER(i2c_dw);
 
 static inline uint32_t get_regs(const struct device *dev)
 {
+#ifdef CONFIG_SOC_CYCLONE5  /* adding an exception for Cyclone V SoC FPGA SoC */
```
why?

mevalver replied:
This is an interesting one...
The approach used in this driver was to use the function DEVICE_MMIO_GET, which wasn't working for us.

This function was defined in include/zephyr/sys/device_mmio.h... And as you can see it calls DEVICE_MMIO_RAM_PTR
which is defined as (mm_reg_t *)((device)->data), hence, it doesn't return the register base address as we want.

I'm open to suggestions, but we determined that this was the most evident way to get this.

carlocaione replied:
Something is fishy here. Do you have `CONFIG_MMU` or `CONFIG_PCIE` enabled?

mevalver replied:
CONFIG_MMU 1
in the generated autoconf.h

CONFIG_PCIE is not set

carlocaione replied:
Right, so if you have the MMU enabled that means that you definitely do not want to use the physical address to access the I2C because almost definitely `phys != virt`. So this is definitely wrong.

mevalver replied:
but, if I dont use this, my i2c simply does not work, thats why I introduced the line... If I get the base address as I did, it works every time....
If I just use whats already in place (DEVICE_MMIO_GET) it simply does not work

carlocaione replied:
It is possible that your MMU is 1:1 mapping between phys and virt, but this is still wrong. Why `DEVICE_MMIO_GET` is not working? what is the address returned? Is your MMU correctly mapping the address?

We do not want to add quirks to workaround issues :)

mevalver replied:
DEVICE_MMIO_GET returns 0
And yes, I was also doubtful about this fix, because it is very specific for the CycloneV... But when I started debugging, I saw that the code in device_mmio.h is not helping me at all ..., is either adding this line or in some way modify .h files in a similar way

carlocaione replied:
`DEVICE_MMIO_GET` should return the `virt` address mapped by the `DEVICE_MMIO_MAP` macro in `i2c_dw_initialize()`.  `DEVICE_MMIO_MAP` is calling `device_map()` to MMU-map the address. Now, what `virt` address is being returned by the `device_map()` call?

mevalver replied:
Moving the get_regs() after DEVICE_MMIO_MAP seems to do the trick. I'll remove this file from this PR, and created another PR to fix that particular bug!

carlocaione reviewed (commented):
This PR has serious style issues.

mevalver commented on soc/arm/intel_socfpga_std/cyclonev/soc.c:
```diff
@@ -0,0 +1,81 @@
+
+static const struct arm_mmu_region mmu_regions[] = {
+
+    MMU_REGION_FLAT_ENTRY("vectors",
```
Solved

mevalver commented on dts/arm/intel_socfpga_std/socfpga.dtsi:
```diff
@@ -0,0 +1,300 @@
+        interrupt-parent = <&intc>;
+        ranges;
+
+		L2: l2-cache@fffef000 {
```
Solved

mevalver commented on dts/arm/intel_socfpga_std/socfpga_cyclone5.dtsi:
```diff
@@ -0,0 +1,28 @@
+/ {
+    soc {
+        clkmgr@ffd04000 {
+            clocks {
```
Solved

mevalver commented on soc/arm/intel_socfpga_std/cyclonev/Kconfig.soc:
```diff
@@ -0,0 +1,12 @@
+# Adding support to Cyclone V SoC FPGA
+
+choice
+    prompt "Intel SoC FPGA Cyclone5"
```
Solved

carlocaione commented:
> dts: arm: / soc: arm Cyclone V SoC FPGA fixing format

This should not be a separate commit. Please squash it.

mevalver commented on dts/arm/intel_socfpga_std/socfpga.dtsi:
```diff
@@ -0,0 +1,300 @@
+    cpus {
+    #address-cells = <1>;
+    #size-cells = <0>;
+    enable-method = "altr,socfpga-smp";
```
bear with me, this is my first contribution to Zephyr. I'm fixing things

carlocaione replied:
I didn't mean to come out strong on this one :smile:

mevalver replied:
this has been fixed, it is not really needed.

mevalver commented on boards/arm/cyclonev_socdk/cyclonev_socdk.dts:
```diff
@@ -0,0 +1,99 @@
+    /* Configuring Zephyr mandatory devices */
+    chosen {
+            bootargs          = "earlyprintk";
+            stdout-path       = "serial0:115200n8";
```
Removed, thanks for pointing out

mevalver commented on samples/drivers/lcd_cyclonev_socdk/src/commands.h:
```diff
@@ -0,0 +1,92 @@
+
+Insert Prefix 0xFE before executing command
+
+*/
```
this has been fixed

mevalver commented on samples/drivers/lcd_cyclonev_socdk/src/main.c:
```diff
@@ -0,0 +1,106 @@
+    }
+    else {
+        printf("Success!\n");
+    }
```
fixed as well

stephanosio commented on drivers/serial/Kconfig.ns16550:
```diff
@@ -34,6 +34,11 @@ config UART_NS16750
 	  flow control if UART controller is 16750.
 
+config UART_NS16750_EXTFIFO
+	bool "Enable UART features (128-bytes FIFO and auto flow control)"
```
Do not start Kconfig names with "Enable" since that is implied; something like "Extended UART 16750 (128-bytes FIFO and auto flow control)" should do.

Also this needs to be a `choice` since `UART_NS16750` and `UART_NS16750_EXTFIFO` are mutually exclusive.

mevalver replied:
what do you mean by "choice"?

stephanosio replied:
Also now that I think about it, if it has 128-byte FIFO, then it is basically 16950, isn't it?

mevalver replied:
I think that's the case... should I change UART_NS16750_EXTFIFO to UART_NS16950?

stephanosio replied:
> should I change UART_NS16750_EXTFIFO to UART_NS16950?

Yes, that would make more sense, IMO.

mevalver replied:
this has been changed

stephanosio commented on drivers/serial/uart_ns16550.c:
```diff
@@ -470,6 +471,8 @@ static int uart_ns16550_configure(const struct device *dev,
 	if ((INBYTE(IIR(dev)) & IIR_FE) == IIR_FE) {
 #ifdef CONFIG_UART_NS16750
 		dev_data->fifo_size = 64;
+#elif CONFIG_UART_NS16950
```
```suggestion
#elif defined(CONFIG_UART_NS16950)
```

mevalver replied:
Solved

stephanosio commented on drivers/serial/Kconfig.ns16550:
```diff
@@ -28,12 +28,24 @@ config UART_NS16550_DRV_CMD
 	  Says n if not sure.
 
+choice
+	prompt "UART x-bytes FIFO and auto flow control"
```
```suggestion
	prompt "UART type"
```

mevalver replied:
solved

stephanosio commented on drivers/serial/Kconfig.ns16550:
```diff
@@ -28,12 +28,24 @@ config UART_NS16550_DRV_CMD
+	prompt "UART x-bytes FIFO and auto flow control"
+	default UART_NS16750
+	help
+	  Select FIFO lenght for serial UART
```
```suggestion
	  Select UART device type
```

mevalver replied:
solved

stephanosio commented on drivers/serial/Kconfig.ns16550:
```diff
@@ -28,12 +28,24 @@ config UART_NS16550_DRV_CMD
 	  flow control if UART controller is 16750.
 
+config UART_NS16950
+	bool "Extended UART 16750 (128-bytes FIFO ans auto flow)"
```
```suggestion
	bool "UART 16950 (128-bytes FIFO and auto flow control)"
```

mevalver replied:
solved

mevalver commented:
@carlocaione, @nashif @stephanosio, @carlescufi, others what else is needed to approve this PR?

teburd commented:
CI needs to pass all its checks

mevalver commented:
Hi again @teburd, @carlocaione,

As a summary:
- The test of the "west commands" keeps failing because we added a "runner" for our board, named "intel_cyclonev", I think for this to pass, the runner need to be added in the assertion list.

- Compliance checks is flagging the format of the "signature" in some commit messages, but if you see the list of commits, I already "rebased" everything and fix the problems in the signature... I don't know if this is running correctly.

-Twister build (1) and "Build with Clang/LLVM" are flagging some kind of issue in the DTS, but I'm not sure what is exactly the problem. Could you please clarify?

Thanks. E.V.

dcpleung commented:
For the complain about author, go through each commit and amend it with `git commit --amend --author="First Last <email>"`. If you do `git --log --format=full`, it will show you both the author and committer.

mevalver commented:
@teburd, @nashif, @carlescufi, @carlocaione, others;

Is there anything pending to get this approved? I haven't had any answer for some days now.

Thanks.
E.V

mevalver reviewed (commented):
This is already fixed

stephanosio reviewed (dismissed):
As noted in https://github.com/zephyrproject-rtos/zephyr/pull/45133#issuecomment-1112391705, please drop/squash commits like:

* [modify: format: adding newlines at the end of the files](https://github.com/zephyrproject-rtos/zephyr/pull/45133/commits/9793cfdaed9764b8783f91b39d1c300ce16940aa)
* boards: arm: Modifying configuration files for Cyclone V SoC DevKit (this fixes some whitespace issues introduced in one of the commits in this PR)
* drivers: serial Removinf unnecessary defines from ns16550 (removing something that was added in this PR ... just do not add it in the first place)
* soc: arm: Configurtaion file for the SoC (commit title does not reflect what it does)
* drivers: serial: adding support for 16950 (contains a lot of unrelated questionable changes, also fixing something that is introduced in this PR)

Please do not fix something you added in this PR in additional commits, fix the original commit instead.

Also there are some commits that contain a lot of unrelated changes.

mevalver committed:
drivers: usb: Adding support to GIC_V1 in dc_dw USB driver

Adding support for the GIC_V1 to the dc_dw USB driver
to be used by Cyclone V SoC FPGA Development Kit

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

mevalver commented:
Commits are organized now!

~E.V.

teburd commented:
The commit set looks quite nice now. I believe the the CI check will still fail due to the west test runner check that looks at the set of available test runners.

```
scripts/west_commands/tests/test_imports.py
```

Should _probably_ have the cyclonev runner added to the expected set to get the import test passing that CI fails on.

teburd commented:
The remaining CI issue looks to be a Kconfig issue with the counter driver being used which probably should depend on but not select I2C, as the devicetree test wants I2C=n

Patch to fix this...

```
--- a/drivers/counter/Kconfig.maxim_ds3231
+++ b/drivers/counter/Kconfig.maxim_ds3231
@@ -5,7 +5,7 @@
 
 config COUNTER_MAXIM_DS3231
 	bool "Maxim DS3231 RTC/TCXO"
-	select I2C
+	depends on I2C
 	select POLL
 	help
 	  Enable counter driver based on Maxim DS3231 I2C device.
```

mevalver committed:
drivers: counter: Setting I2C as depends in DS3231 config file

When using DS3231 counter, its enablement should depend on I2C

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

teburd commented on boards/arm/cyclonev_socdk/cyclonev_socdk.dts:
```diff
@@ -0,0 +1,88 @@
+
+	leds {
+		 compatible = "gpio-leds";
+		 hps3: hps_led_3 {
```
nit: Space after tab here and below for the leds section

mevalver replied:
this is fixed

teburd reviewed (dismissed):
Looks really nice now, one last format nit that needs a tweak, pre-emptive +1 for the good feels

carlocaione commented on drivers/usb/device/usb_dc_dw.c:
```diff
@@ -719,7 +719,12 @@ int usb_dc_attach(void)
 	IRQ_CONNECT(DT_INST_IRQN(0),
 		    DT_INST_IRQ(0, priority),
 		    usb_dw_isr_handler, 0,
+#ifdef CONFIG_GIC_V1
+		    DT_INST_IRQ(0, type));
+#else
 		    DT_INST_IRQ(0, sense));
+#endif
+
```
Can you explain this change?

mevalver replied:
Hi @carlocaione,

First of all the use of DT_INST_IRQ (0, sense), is very specific to snps,designware-intc... You can see this in the biding documentation  https://docs.zephyrproject.org/latest/build/dts/api/bindings/interrupt-controller/snps,designware-intc.html.... so, the USB driver shouldn't have that definition when using it with arm,gic interrupt controller.

In the modification, we just decided to configure the "type" cell in the binding. Instead of the "sense" cell... This does not mean that both of them are interchangeable.

I believe, we can delete "DT_INST_IRQ(0, type)" from the file, as is not strictly necessary. But the use of "DT_INST_IRQ(0, sense)" should be restricted only when using the Synopsys interrupt controller.

~E.V.

carlocaione replied:
uhm, this sucks. @mbolivar-nordic do you have any smart solution about this? The change is fine but it is definitely not scalable if we decide to use this IP with different interrupt controllers with different cell names.

stephanosio replied:
@mevalver So, based on your comment, the logic should be "if its interrupt parent is `snps,designware-intc,` use `sense`; else, use `type`."

mevalver replied:
I think that could be the case, as "sense" is specific cell-name to snps,designware-intc

carlocaione reviewed (approved):
LGTM but this should be rebased and merged after #45253

stephanosio reviewed (approved):
Looks ok overall

Dino-Li commented:
@mevalver 
The change of Kconfig.ns16550 breaks NS16550 UART device.
The newly added "chioce" select UART_NS16750 by default.
But we should not enable UART_NS16750 or UART_NS16950 on NS16550  because Its fifo size is 16 bytes.
Can the "choice" be removed?

teburd commented:
> @mevalver The change of Kconfig.ns16550 breaks NS16550 UART device. The newly added "chioce" select UART_NS16750 by default. But we should not enable UART_NS16750 or UART_NS16950 on NS16550 because Its fifo size is 16 bytes. Can the "choice" be removed?

Please file an issue and assign to me, tagging @mevalver
